### ISSUE TITLE
Sockets.get(peer|sock)name on closed TCPSocket segfaults

### ISSUE THREAD

**Issue by fredrikekre** [JuliaLang/julia#40993]
Could probably be an error instead?

`sock.handle` is a null pointer here
https://github.com/JuliaLang/julia/blob/5dbf45a2997acd26e0f72abaa320c9d0d54956a1/stdlib/Sockets/src/Sockets.jl#L817 and here https://github.com/JuliaLang/julia/blob/5dbf45a2997acd26e0f72abaa320c9d0d54956a1/stdlib/Sockets/src/Sockets.jl#L821

Would `peer = isopen(sock) ? getpeername(sock) : nothing` be a safe alternative to avoid this segfault (or error if this issue is fixed)?

**Comment by vtjnash** [comment]
Normally we'd call `check_open(sock)` somewhere near the top of the function, and let that throw an exception if the state is invalid for access

**Comment by cljord** [comment]
Hi @vtjnash @fredrikekre 

I'm pretty new to contributing to open-source and Julia, but it sounds like the solution is really just adding the check_open(sock) check to the top of the function (I took a look at check_open and it throws an error itself). If I understood that correctly, I'd like to take this issue.

**Comment by cljord** [comment]
Hi again,
so I added the check_open(sock) to the beginning of the function, but now the tests fail. I've poked around a little and the buildbot always has a:

nested task error: ArgumentError: TCPSocket(RawFD(19) init, 0 bytes waiting) is not initialized

from runtests.jl, starting at line 93.

Sorry for probably adding more work than I solved, but I'm in a bit over my head with this large a codebase and I can't figure it out. My pull request is #41000 for a closer look at the buildbot.
